src/librustc/middle/wf.rs

```diff
@@ -0,0 +1,546 @@
+                            match traits::trait_ref_for_builtin_bound(this.infcx.tcx,
+                                                                      ty::BoundSized,
+                                                                      subty) {
+                                Ok(trait_ref) => {
+                                    this.out.push(
+                                        traits::Obligation::new(cause,
+                                                                trait_ref.to_predicate()));
+                                }
+                                Err(ErrorReported) => { }
+                            }
+                        }
+                    })
+                }
+
+                ty::TyBox(_) |
+                ty::TyTuple(_) |
+                ty::TyRawPtr(_) => {
+                    // simple cases that are WF if their type args are WF
+                }
+
+                ty::TyProjection(data) => {
+                    subtys.skip_current_subtree(); // subtree handled by compute_projection
+                    self.compute_projection(data);
+                }
+
+                ty::TyEnum(def, substs) |
+                ty::TyStruct(def, substs) => {
+                    // WfNominalType
+                    let obligations = self.nominal_obligations(def.did, substs);
+                    self.out.extend(obligations);
+                }
+
+                ty::TyRef(r, mt) => {
+                    // WfReference
+                    if !r.has_escaping_regions() && !mt.ty.has_escaping_regions() {
+                        let cause = self.cause(traits::ReferenceOutlivesReferent(ty));
+                        self.out.push(
+                            traits::Obligation::new(
+                                cause,
+                                ty::Predicate::TypeOutlives(
+                                    ty::Binder(
+                                        ty::OutlivesPredicate(mt.ty, *r)))));
+                    }
+                }
+
+                ty::TyClosure(..) => {
+                    // the types in a closure are always the types of
+                    // local variables (or possibly references to local
+                    // variables), which are separately checked w/r/t
+                    // WFedness.
+                }
+
+                ty::TyBareFn(..) => {
+                    // process the bound types; because the old implicator
+                    // did not do this, go into RFC1214 mode.
+                    subtys.skip_current_subtree();
+                    self.compute_rfc1214(ty);
+                }
+
+                ty::TyTrait(ref data) => {
+                    // WfObject
+                    //
+                    // Here, we defer WF checking due to higher-ranked
+                    // regions. This is perhaps not ideal.
+                    self.from_object_ty(ty, data);
+
+                    // FIXME(#27579) RFC also considers adding trait
```

Comment:
@arielb1 

> Couldn't we check the obligations with Self being the object type?

Interesting thought.  I'll have to think it over. It would certainly be nice if so. It may interact with the restrictions on object safety and #27675.

> I'm not sure how important is this, because (am I wrong here?) non-supertrait predicates are never elaborated and therefore can be safely ignored

This is true but I have considered changing this (https://github.com/rust-lang/rust/issues/20671).